### PR TITLE
[#418] Propagate prefill information from deployment chain to pool co…

### PR DESCRIPTION
**Pull request by maeste**
…re deployers major task and add test cases

### CODE REVIEW

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 1]
```diff
@@ -62,7 +63,14 @@
    
    /** The semaphore */
    protected Semaphore semaphore;
-   
+
+   private boolean shouldPrefill;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Remove

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 2]
```diff
@@ -62,7 +63,14 @@
...
+
+   private Credential prefillCredential;
+
+   /** Shutdown */
+   private final AtomicBoolean shutdown = new AtomicBoolean(false);
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Remove

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 3]
```diff
@@ -75,6 +83,7 @@ public AbstractPool(ConnectionManager cm, PoolConfiguration pc)
       this.pools = new ConcurrentHashMap<Credential, ManagedConnectionPool>();
       this.transactionMap = new ConcurrentHashMap<Object, Map<ManagedConnectionPool, ConnectionListener>>();
       this.semaphore = new Semaphore(poolConfiguration.getMaxSize());
+      this.shouldPrefill = pc.isPrefill(); // || pc.isStrictMin();
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Remove

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 4]
```diff
@@ -178,6 +165,34 @@ public ConnectionListener getConnectionListener(Credential credential)
       return cl;
    }
 
+   private ManagedConnectionPool getManagedConnectionPool(Credential credential)
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Unrelated change - or add JavaDoc to method

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 5]
```diff
@@ -204,6 +219,7 @@ public boolean isFull()
     */
    public synchronized void shutdown()
    {
+      shutdown.set(true);
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Remove

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 6]
```diff
@@ -211,6 +227,15 @@ public synchronized void shutdown()
    }
 
    /**
+    * {@inheritDoc}
+    */
+   @Override
+   public boolean isShutdownInProgress()
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: boolean isShutdown() delegates to AbstractConnectionManager

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 7]
```diff
@@ -397,4 +422,33 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
     */
    protected abstract ManagedConnectionPool createManagedConnectionPool(Credential credential);
+
+   /**
+    * Prefill the connection pool
+    *
+    * @param credential the pool credential
+    */
+   @Override
+   public void prefill(Credential credential)
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Remove 'credential' from signature

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 8]
```diff
@@ -397,4 +422,33 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+   {
+
+      this.prefillCredential = credential;
+      if (shouldPrefill)
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Check PoolConfiguration directly

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 9]
```diff
@@ -397,4 +422,33 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+      {
+         ManagedConnectionPool mcp = getManagedConnectionPool(credential);
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: getPrefillCredential()

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 10]
```diff
@@ -397,4 +422,33 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+      }
+   }
+
+   /**
+    * Get prefill credential
+    * @return credential used to prefill
+    */
+   public Credential getPrefillCredential()
+   {
+      return this.prefillCredential;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Construct prefillCredential if null

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/PoolFiller.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+   private static final PoolFiller FILLER = new PoolFiller();
+
+   /** Pools list */
+   private final LinkedList<FillRequest> pools = new LinkedList<FillRequest>();
+
+   /** Filler thread */
+   private final Thread fillerThread;
+
+   /** Thread name */
+   private static final String THREAD_FILLER_NAME = "JCA PoolFiller";
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/PoolFiller.java]: JCA -> IronJacamar

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 12]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
       }
       listeners.clear();
    }
+
+   /**
+    * Prefill
+    */
+   @Override public void prefill()
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: '@Override' misplaced

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 13]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+   {
+      if (this.credential.equals(pool.getPrefillCredential()))
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: Add more guards -- isPrefill() + MinPooSize() > 0

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 14]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+      {
+         PoolFiller.fillPool(new FillRequest(this, pool.getConfiguration().getInitialSize()));
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: This should be getMinSize() -- the InitialPoolSize() should be used in the constructor

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 15]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+    * Fill to
+    *
+    * @param size The size
+    */
+   public void fillTo(int size)
+   {
+      if (size <= 0)
+         return;
+
+      if (!pool.getConfiguration().isPrefill())
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: This shouldn't be here

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 16]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+         return;
+
+      if (!credential.equals(pool.getPrefillCredential()))
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: This shouldn't be here

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 17]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+      //TODO: trace and debug here
+
+      while (!pool.isFull())
+      {
+         // Get a permit - avoids a race when the pool is nearly full
+         // Also avoids unnessary fill checking when all connections are checked out
+         try
+         {
+            //TODO:statistics
+            if (pool.getRequestSemaphore()
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: This shouldn't be in DefaultMCP

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 18]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+                  }
+
+                  // Create a connection to fill the pool
+                  try
+                  {
+                     ConnectionListener cl = pool.createConnectionListener(credential);
+
+                     //TODO:Trace
+                     boolean added = false;
+                     synchronized (listeners)
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: It is a concurrent data structure

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 19]
```diff
@@ -165,4 +170,99 @@ else if (cl.getState() == DESTROY)
...
+                     }
+                  }
+                  catch (ResourceException re)
+                  {
+                     return;
+                  }
+               }
+               finally
+               {
+                  pool.getRequestSemaphore().release();
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: Remove

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java [hunk 20]
```diff
@@ -127,4 +128,13 @@ protected ManagedConnectionPool createManagedConnectionPool(Credential credentia
    {
       return new DefaultManagedConnectionPool(this, credential);
    }
+
+   /**
+    * Get the request semaphore
+    * @return The value
+    */
+   Semaphore getRequestSemaphore()
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java]: Nope

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java [hunk 21]
```diff
@@ -178,4 +187,102 @@ else if (cl.getState() == DESTROY)
       }
       listeners.clear();
    }
+
+   /**
+    * Prefill
+    */
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java]: Similar comments

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java [hunk 22]
```diff
@@ -178,4 +187,102 @@ else if (cl.getState() == DESTROY)
...
+      {
+         PoolFiller.fillPool(new FillRequest(this, pool.getConfiguration().getInitialSize()));
+      }
+   }
+
+   /**
+    * Fill to
+    *
+    * @param size The size
+    */
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java]: Similar comments

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StablePool.java [hunk 23]
```diff
@@ -175,4 +175,5 @@ Semaphore getRequestSemaphore()
    {
       return requestSemaphore;
    }
+
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StablePool.java]: Unneeded

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 24]
```diff
@@ -35,6 +35,7 @@
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: This is unused

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 25]
```diff
@@ -397,4 +422,49 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+   public void prefill()
+   {
+
+      if (poolConfiguration.isPrefill())
+      {
+         ManagedConnectionPool mcp = pools.get(getPrefillCredential());
+
+         if (mcp == null)
+         {
+            getManagedConnectionPool(getPrefillCredential());
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Missing assignment

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 26]
```diff
@@ -397,4 +422,49 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+         } else
+         {
+            mcp.prefill();
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Should be outside the 'if'

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java [hunk 27]
```diff
@@ -31,6 +31,7 @@
 import org.ironjacamar.core.connectionmanager.pool.AbstractPool;
 import org.ironjacamar.core.connectionmanager.pool.ManagedConnectionPool;
 
+import java.util.concurrent.Semaphore;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java]: Unused

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java [hunk 28]
```diff
@@ -127,4 +128,5 @@ protected ManagedConnectionPool createManagedConnectionPool(Credential credentia
    {
       return new DefaultManagedConnectionPool(this, credential);
    }
+
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultPool.java]: White space

File: testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/PrefillCriTestCase.java [hunk 29]
```diff
@@ -0,0 +1,650 @@
...
+      assertNotNull(mcps);
+      assertEquals(1, mcps.size());
+
+
+      UnifiedSecurityConnection c =  noTxCf.getConnection("user", "pwd");
+      assertNotNull(c);
+      assertEquals("user", c.getUserName());
+      assertEquals("pwd", c.getPassword());
+
+      UnifiedSecurityConnection c1 = noTxCf.getConnection("user", "pwd");
```
Comment (jesperpedersen) [inline testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/PrefillCriTestCase.java]: We only need to test connections that are different than the prefill credentials

File: testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/PrefillCriTestCase.java [hunk 30]
```diff
@@ -0,0 +1,650 @@
...
+
+      assertEquals(2, mcps.size());
+      Iterator<ManagedConnectionPool> mcpsIter = mcps.values().iterator();
+      ManagedConnectionPool mcp = mcpsIter.next();
+      assertNotNull(mcp);
+
+      ConcurrentLinkedDeque<ConnectionListener> listeners = (ConcurrentLinkedDeque<ConnectionListener>) TestUtils
+            .extract(mcp, "listeners");
+      assertNotNull(listeners);
+      assertEquals(10, listeners.size());
```
Comment (jesperpedersen) [inline testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/PrefillCriTestCase.java]: That won't work on slow machines -- you need to use a ```void untilSize(Collection<>, int size, int timeout, TimeUnit unit)``` method

File: testsuite/src/test/java/org/ironjacamar/rars/ResourceAdapterFactory.java [hunk 31]
```diff
@@ -345,7 +345,7 @@ public static ResourceAdapterArchive createUnifiedSecurityRar()
     * @return The resource adapter descriptor
     */
    public static ResourceAdaptersDescriptor createUnifiedSecurityDeployment(String bc, String securityDomain,
```
Comment (jesperpedersen) [inline testsuite/src/test/java/org/ironjacamar/rars/ResourceAdapterFactory.java]: Maybe do some overloaded methods, so you don't have to pass in all parameters in all tests

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java [hunk 32]
```diff
@@ -397,4 +422,49 @@ else if (txCM.getTransactionIntegration().isConnectableResource(mc))
...
+    *
+    * @param credential the pool credential
+    */
+   @Override
+   public void prefill()
+   {
+
+      if (poolConfiguration.isPrefill())
+      {
+         ManagedConnectionPool mcp = pools.get(getPrefillCredential());
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: You can just call ```mcp = getManagedConnectionPool(getPrefillCredential());``` directly -- it has the pools.get()

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java (hunk 26 above)
Comment (maeste) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: well it's intentional. prefill w/ initial-pool-size is triggered by mcp creation in getManagedConnectionPool(getPrefillCredential())). Explicit call to prefill is needed only on  already present mcp

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java (hunk 25 above)
Comment (maeste) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: it's just to trigger mcp creation. I'm not assigning it because I don't need to do anything else on just created mcp

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java (hunk 32 above)
Comment (maeste) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: yup it has, but there is no way to understand if it is an already present mcp or created at the moement of the call. See comment below

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java (hunk 26 above)
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/AbstractPool.java]: Ok, lets keep it like that, but add comments
```
if (mcp == null)
{
   // Trigger the initial-pool-size prefill by creating the ManagedConnectionPool
  mcp = ...
}
else
{
    // Standard prefill request
   mcp.prefill();
}
```
